### PR TITLE
Add/flux framework scheduler

### PR DESCRIPTION
**Pull request by vsoch**
This is the first draft to add the Flux Scheduler to Pavilion2! The Flux way of running jobs isn't nicely matched to slurm so some of these interactions were harder to figure out, but this should be a good start for us to discuss. Thanks for all the great help @pflarr !

### CODE REVIEW

File: lib/pavilion/schedulers/plugins/flux.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+flux_states = [
+    "DEPEND",
+    "SCHED",
+    "RUN",
+    "CLEANUP",
+    "COMPLETED",
+    "FAILED",
+    "CANCELED",
+    "TIMEOUT",
+]
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/flux.py:35]: The equivalent in the Slurm scheduler was about figuring out the state of each node, but these look like job states. If that's what you're using them for, that's fine.

File: lib/pavilion/schedulers/plugins/flux.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        }
+
+        return elems, validators, defaults
+
+    @classmethod
+    def parse_node_list(cls, node_list) -> NodeList:
+        """
+        Convert a list of strings into a list of nodes
+        """
+        return NodeList(node_list)
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/flux.py:81]: You can just get rid of this, and do this inline. A lot of these functions are Slurm specific. Check the Advanced and base scheduler classes for the methods that must be overridden (they'll all raise NotImplementedError).

File: lib/pavilion/schedulers/plugins/flux.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        Get a flux resource list
+        """
+
+        rpc = flux.resource.list.resource_list(flux.Flux())
+        listing = rpc.get()
+
+        nodelist = [str(node) for node in listing.free.nodelist]
+        # Flux doesn't currently support reservations
+        extra = {"reservations": {}, "nodes_listing": listing}
+        return nodelist, extra
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/flux.py:107]: The data this returns is entirely dependent on the Scheduler. No need to still return anything about reservations.

File: lib/pavilion/schedulers/plugins/flux.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    def _transform_raw_node_data(self, sched_config, node_data, extra) -> NodeInfo:
+        """
+        Translate the gathered data into a NodeInfo dict.
+        """
+        listing = extra["nodes_listing"]
+        del extra["nodes_listing"]
+        node_info = NodeInfo({})
+        node_info["name"] = node_data
+        node_info["cpus"] = listing.free.ncores
+        node_info["states"] = None
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/flux.py:118]: See the superclass for the list of required keys. Anything else is scheduler specific, including 'states'.

File: lib/pavilion/schedulers/plugins/flux.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        node_info["available"] = node_data
+        return node_info
+
+    def _filter_custom(
+        self, sched_config: dict, node_name: str, node: NodeInfo
+    ) -> Union[str, None]:
+        """
+        Filter nodes by features. Flux doesn't have features.
+        """
+        return None
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/flux.py:129]: The default does nothing, so you can scrap this.

File: lib/pavilion/schedulers/plugins/flux.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+    def _available(self) -> bool:
+        """
+        Ensure we can import and talk to flux.
+        """
+        return flux is not None
+
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/flux.py:136]: Everything from here down looks good.

File: lib/pavilion/schedulers/plugins/flux.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        Translate the gathered data into a NodeInfo dict.
+        """
+        listing = extra["nodes_listing"]
+        del extra["nodes_listing"]
+        node_info = NodeInfo({})
+        node_info["name"] = node_data
+        node_info["cpus"] = listing.free.ncores
+        node_info["states"] = None
+        node_info["up"] = node_data
+        node_info["available"] = node_data
```
Comment (Paul-Ferrell) [inline lib/pavilion/schedulers/plugins/flux.py:120]: These have a very important and specific meaning (from the parent method doc string):

>        # Node Status - (required)
>
>        - up (bool) - Whether the node is up (allocatable).
>        - available (bool) - Whether the node is allocatable and unallocated.

File: lib/pavilion/schedulers/plugins/flux.py (hunk 3 above)
Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py:107]: ah gotcha.

File: lib/pavilion/schedulers/plugins/flux.py (hunk 4 above)
Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py:118]: Where is that? I did a grep and only found a test file dummy.py

Comment (Paul-Ferrell) [reply lib/pavilion/schedulers/plugins/flux.py:118]: schedulers.advanced.py

File: lib/pavilion/schedulers/plugins/flux.py (hunk 7 above)
Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py:120]: So the sets overlap?

Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py]: okay we actually have both free and up lists to match nicely to these two!

File: lib/pavilion/schedulers/plugins/flux.py (hunk 4 above)
Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py]: Thanks! Okay so the list is:

- name: we have this
- status (up/available): now are correctly specified (bool)
- cpus: we have this
- mem: I didn't see this information with the listing

And these are slurm specific I think.

- partitions (List) - The cluster partitions on which the node resides.
- reservations (List) - List of reservations to which the node belongs.
- features (List[str]) - A list of feature tags that differentiate nodes,

File: lib/pavilion/schedulers/plugins/flux.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        if job_info["sys_name"] != sys_name:
+            return TestStatusInfo(
+                STATES.SCHEDULED,
+                "Job started on a different cluster ({}).".format(sys_name),
+            )
+
+        listing = flux.job.list.JobList(flux.Flux(), ids=[job_info["jobid"]])
+        jobs = listing.jobs()
+
+        if not listing:
```
Comment (grondo) [inline lib/pavilion/schedulers/plugins/flux.py:185]: Should this be `if not jobs`?

**Review by grondo — COMMENTED** [review]
Just adding some comments on use of the Flux APIs after a cursory inspection.

File: lib/pavilion/schedulers/plugins/flux.py (hunk 4 above)
Comment (Paul-Ferrell) [reply lib/pavilion/schedulers/plugins/flux.py]: partitions and reservations are general, but optional.

File: lib/pavilion/schedulers/plugins/flux.py (hunk 8 above)
Comment (Paul-Ferrell) [reply lib/pavilion/schedulers/plugins/flux.py:185]: Style note - Pavilion has a separate job object, so we should probably be disambiguating them.  Maybe call them `flux_jobs`?

File: lib/pavilion/schedulers/plugins/flux.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+                note="Could not find job {}, must have finished".format(job_info["id"]),
+                when=time.time(),
+            )
+
+        # State list is here
+        # https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_21.html#state-diagram
+        # DEPEND
+        job = jobs[0]
+
+        if job.state == "COMPLETED":
```
[the remark this answers is not in the document]
Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py:197]: ah ok, so I should be checking status here!

File: lib/pavilion/schedulers/plugins/flux.py (hunk 8 above)
Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py]: yes it should! I renamed / debugged this a bunch of times and forgot to update the name here, should be fixed now.

**Review by jameshcorbett — COMMENTED** [review]
One insignificant comment below.

File: lib/pavilion/schedulers/plugins/flux.py [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+        # This doesn't seem to be used?
+        fluxjob.cwd = str(job.tests_path)
+        fluxjob.environment = dict(os.environ)
+
+        # This submits without waiting
+        flux_future = flux.job.submit_async(flux.Flux(), fluxjob)
+
+        # This blocks until the RPC is complete so we will get an ID
+        jobid = flux_future.get_id()
```
[the remark this answers is not in the document]
Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py:150]: ah! so submit_async doesn't wait for the id, but I'm asking it to wait with get_id? So if I just use submit that does two in one?

[the remark this answers is not in the document]
Comment (jameshcorbett) [reply lib/pavilion/schedulers/plugins/flux.py:150]: Yeah. Sorry for not answering earlier.

[the remark this answers is not in the document]
Comment (jameshcorbett) [reply lib/pavilion/schedulers/plugins/flux.py:150]: [The code for `submit` is actually just a wrapper around submit_async](https://github.com/flux-framework/flux-core/blob/99ecf1da6ebcacf4e53ced93f0088c039de1d12e/src/bindings/python/flux/job/submit.py#L103)

[the remark this answers is not in the document]
Comment (vsoch) [reply lib/pavilion/schedulers/plugins/flux.py:150]: no worries! I think I realized this when I was working on the asyncio stuff - apologies for asking I should have just looked, still familiarizing with the python code.

**Comment by Paul-Ferrell** [comment]
Flux scheduler finished an merged in #674

**Comment by vsoch** [comment]
Nice work @Paul-Ferrell so happy to see that merged! :partying_face:

**Pull request closed without merging**